### DISCUSSION TITLE
NextJS, env vars, the CI/CD and the build stages....

### DISCUSSION THREAD

**Discussion by exocode** [vercel/next.js#30691] [Help]
Hi there, 

I am struggling with the concept of hard coded client ENV vars. Many apps we build are simply built into a container with Docker.
We then deploy them to a - let's say - "staging" environment.  In that environment exists staging env vars for APIs and domains. If all of our tests are passing, we handle over that image to "Q/A" or "Production".
In that new stage we need different env-vars values (eg: the "real" API endpoints, and a different domain, like www). My approach till now was to build one single image and simply provide different env-vars in that different stages.

<img width="1061" alt="Bildschirmfoto 2021-10-31 um 12 08 54" src="https://user-images.githubusercontent.com/1597621/139580041-ea648807-5295-4990-a2b9-38694634bd94.png">

**Now to my concern:**
Isn't it the purpose of an built image, to handle over the exact image to any stage,  to save time/money and for consistentcy and simply swap the env-var-values? 

In NextJS I have to rebuild the whole image over and over for each building stage, because of that "baked in env-vars". They differ in each stage and are only valid for that specific stage. With that approach a regular build looks like this: 

1. Build **test image**
2. Do some tests (specs)
3. Build a **staging image**
4. Do some tests
5. Build a **Q/A image**
6. Do some tests
7. Build the **production image**
8. publish the image

In that example, 4 times the same image is built over and over, except the env-vars. which is not only time consuming, it's also costly ($$$).

In my case I use Gitlab [AutoDevOps](https://docs.gitlab.com/ee/topics/autodevops/) which is quite handy and battle tested and best practice. But for NextJS it must be adapted/expanded quite heavily to just follow the pattern above. At the end my app is deployed in a Kubernetes cluster (not in Vercel)

How do you handle that? Do I have a misunderstanding of that whole env-var concept of NextJS?

Thank you in advance

Scheme of my "AutoDevOps" CD/CD:

**Comment by patjakubik** [comment]
Relevant https://github.com/vercel/next.js/discussions/22243
